gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,146 @@
+        super().__init__()
+
+        self.interface = interface
+
+        self.geoclue_simple = None
+        self.notify_location_handle = None
+        self.notify_active_listener = None
+        self.in_flatpak = None
+
+        def new_finish(_, async_result):
+            try:
+                self.geoclue_simple = Geoclue.Simple.new_finish(async_result)
+            except GLib.Error as e:
+                if e.matches(Gio.DBusError, Gio.DBusError.ACCESS_DENIED):
+                    # TODO: In practice, I cannot get this to occur. Maybe it is KDE's portal
+                    # implementation, but when I deny location permissions in the prompt,
+                    # the failure is a generic DBusError.FAILED, not a permission denied
+                    # failure.
+                    self.props.state = States.DENIED
+                else:
+                    self.props.state = States.FAILED
+                    print("Failed to get location", e.message, e.code)
+
+                return
+            else:
+                self.props.state = States.READY
+
+            client = self.geoclue_simple.get_client()
+            # Geoclue docs indicate a client proxy is not used in Flatpak
+            # https://lazka.github.io/pgi-docs/#Geoclue-2.0/classes/Simple.html#Geoclue.Simple.props.client
+            self.in_flatpak = client is None
+
+            if client:
+
+                def notify_client_active(*args):
+                    if not self.geoclue_simple.get_client().props.active:
+                        # If the client explicitly becomes inactive, this indicates
+                        # a failure to retrieve the location
+                        self.props.state = States.DENIED
+
+                self.notify_active_listener = client.connect(
+                    "notify::active", notify_client_active
+                )
+
+        Geoclue.Simple.new(
+            self.interface.app.app_id, Geoclue.AccuracyLevel.EXACT, None, new_finish
+        )
+
+    @property
+    def can_get_location(self):
+        return self.props.state in (States.READY, States.MONITORING)
+
+    def has_permission(self):
+        # TODO: Think this through a bit more
+        if self.props.state == States.STARTING:
+            # Cannot know whether permission request is needed...
+            # False is safe?
+            return False
+
+        if self.in_flatpak and not self.can_get_location:
+            # In flatpak, explicit permission required
+            return False
+
+        # Otherwise, no explicit permission required
+        return True
+
+    def request_permission(self, result):
+        if self.can_get_location:
+            result.set_result(True)
+        elif self.props.state == States.STARTING:
+
+            def wait_for_client(*args):
+                result.set_result(self.can_get_location)
+                self.disconnect(listener_handle)
+
+            listener_handle = self.connect("notify::state", wait_for_client)
```

Comment:
Here's the "magic" that inheriting from `GObject.Object` enables :tada:. I tried to think of other ways to do this, but they were all nasty and over-complicated. At the end of the day, requesting permission needs to be aware of the asynchronous nature of the Geoclue client startup, and have a way to schedule itself until after the client is available if this is called early.

Reply:
Looks elegant to me. I might question whether it would be worth making the "notification" object an *attribute* of the impl, rather than actually being the impl - but the notification mechanism will ultimately be the same, so maybe the extra object isn't worth it if the impl can do double duty.

The other option would be to subclass the GeoClue client class to add this "state" property, so everything can be done on a single "native" object.